src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
+    }
+
+    /* Check that we are allowed to send to this cap */
+    if (unlikely(!cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+    tcb_t *dest;
+
+    if (ntfnState == NtfnState_Waiting) {
+        /* get the destination thread */
+        dest = TCB_PTR(notification_ptr_get_ntfnQueue_head(ntfnPtr));
+    } else {
+        /* get the bound tcb */
+        dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+    }
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Signal to higher prio thread is NOT fastpathed*/
+            if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+             * a SC  or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+            }
+
+            /*  Point of no return */
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+            endpoint_ptr_set_epQueue_head_np(ep_ptr, TCB_REF(dest->tcbEPNext));
+            if (unlikely(dest->tcbEPNext)) {
+                dest->tcbEPNext->tcbEPPrev = NULL;
+            } else {
+                endpoint_ptr_mset_epQueue_tail_state(ep_ptr, 0, EPState_Idle);
+            }
+
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            if (!dest->tcbSchedContext) {
+                schedContext_donate(sc, dest);
+            }
+            assert(dest->tcbSchedContext);
+
+            /* Enqueue signalled but continue executing signaller */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
```

Comment:
In both cases when the new thread is set to Running, there needs to be a check for `sc_sporadic(dest->tcbSchedContext)` and if it is a follow call to `refill_unblock_check()`

Reply:
> > As far as I can tell we should have equivalence with the slow path now (but I could have missed something, of course), and we're not performing any state changes before we abort into the slow path.
> 
> Has there been any discussion about the lack of checkBudgetRestart() on this fastpath? It's not present on any of the other fastpaths either which may be an existing oversight?

The initial design avoided fastpathing any case where an SC would be charged at all in the fastpath, and thereby avoids any timer reprograms or handling of budget exhaustion, ruling out a large amount of branches. This assumes you're always staying on the same SC.

Reply:
I'm assuming that the goal of this PR is to fastpath the cases where there is a thread blocked on receiving either on the notification or on an Endpoint but with the notification bound to the TCB, and the thread is of equal or lower priority of the signalling thread. In any case where the other thread is made unblocked by the signal, we need to ensure that the MCS operations are handled. As refill_unblock_check() cannot be deferred and must occur when the SC for the signalled thread becomes active it can't be skipped here, otherwise it wouldn't be run at all.

Reply:
if that's what we're fastpathing then yeah, all bets are off wrt to budgets and we can't assume it's all fine. 

I would add a case to bail immediately if we see our sc's doesn't meet the conditions for a simple switch (e.g the SC we switch from has enough budget + refills to charge and stay in the scheduler queue, and the SC we are switching to).